### ISSUE TITLE
generic type arguments set as types of optional properties become "any" when not set

### ISSUE THREAD

**Issue by captaincaius** [microsoft/TypeScript#37235]
**TypeScript Version:**  3.8.3

**Search Terms:**

optional property generic

**Expected behavior:**

type variable inferred as undefined when optional property is missing

**Actual behavior:**

**Related Issues:**

none

**Code**
```ts
type AnySanitizer = (unsafeParams: any) => any;
type OptionalSanitizer = AnySanitizer | undefined;

interface MaybeSanitizing<TSanitizer extends OptionalSanitizer> {
    sanitizeParams?: TSanitizer;
    params: TSanitizer extends AnySanitizer ? ReturnType<TSanitizer> : null;
}

class SomeHT1 {
    // The case where the next line is uncommented correctly complains.
    //sanitizeParams(paramsIn: any) { return "yo";}
    params = 3;
}

function doSomethin<SaniType extends OptionalSanitizer>(someClass: MaybeSanitizing<SaniType>) {
    return someClass.params;
}

const h = new SomeHT1();

const yo = doSomethin(h);
// EXPECTED: Error because params needs to be null
// ACTUAL: No error

```

<details><summary><b>Output</b></summary>

```ts
"use strict";
class SomeHT1 {
    constructor() {
        // The case where the next line is uncommented correctly complains.
        //sanitizeParams(paramsIn: any) { return "yo";}
        this.params = 3;
    }
}
function doSomethin(someClass) {
    return someClass.params;
}
const h = new SomeHT1();
const yo = doSomethin(h);
// EXPECTED: Error because params needs to be null
// ACTUAL: No error

```

</details>

<details><summary><b>Compiler Options</b></summary>

```json
{
  "compilerOptions": {
    "noImplicitAny": true,
    "strictNullChecks": true,
    "strictFunctionTypes": true,
    "strictPropertyInitialization": true,
    "strictBindCallApply": true,
    "noImplicitThis": true,
    "noImplicitReturns": true,
    "useDefineForClassFields": false,
    "alwaysStrict": true,
    "allowUnreachableCode": false,
    "allowUnusedLabels": false,
    "downlevelIteration": false,
    "noEmitHelpers": false,
    "noLib": false,
    "noStrictGenericChecks": false,
    "noUnusedLocals": false,
    "noUnusedParameters": false,
    "esModuleInterop": true,
    "preserveConstEnums": false,
    "removeComments": false,
    "skipLibCheck": false,
    "checkJs": false,
    "allowJs": false,
    "declaration": true,
    "experimentalDecorators": false,
    "emitDecoratorMetadata": false,
    "target": "ES2017",
    "module": "ESNext"
  }
}
```

</details>

**Playground Link:** [Provided](http://www.typescriptlang.org/play/index.html#code/C4TwDgpgBAggdiAygQzgS2GgXhATlAXigAoBXOAZ2QDMIAFZXZAWwoC4pUQBKQgPk4IA3AFgAUKEhQA8mEwB7OMgA2KdJhz4i8JKgzY8UAD5RyAEwjU0cCGdFjx14HmrIAxtACyyEACMIavpY1gDmADwAKoEahhAAHs5wZhQycmiKKtEGuAIA3uJQhVBU6gYMTKwA-BxRejG49kVQYIws7FC1pZpQ8YnJsAhZ3ZVQAEoQwKS4cBHgEJFDeAIccKTKyvYAvuLibsrIFCmI8swQABIRAIxQ+WJNAPT3HQAW0G4H0ADur7jQwK9QGwJKDKazQNApchuE6nODOMxQaG4X5uYDKECIk5gfbWCgAOgKRUeJSC9FarGILQqFAAknAOFxeLkoL9JtMoAAiEDyDlCbZ3IpUtqEKAAZi2OzE1ChCjgUDM8mOp3+1jCgVmUl6ECSKVkssydWyfGIFBOEAAwvtDhxvH4AobgnBwuq5nwmYTCqypnLTadLQd8ULWBKHGJoZRgFBniKbJ8oErzldiNx7LtFBRI9yRQqEyq4MRninxI8oABRAAadFL5oipYAIhxS8j5Ph-O9SBRoEGUjZbClgPIoP5AWtlMWnjAawBVGAAGQ4ADlB3hcC3xEA)

**Comment by captaincaius** [comment]
Just a quick note in case it's not clear - my goal in this code is to enforce that the "params" property's type depends on the existence of the sanitizeParams method:  If it exists, it must be the return type of the method.  If it doesn't, it must be null.

There seems to be no way to achieve this.

**Comment by RyanCavanaugh** [comment]
I'm not really clear on what the proposed defect is. Can you add more examples of what you're trying to do?

**Comment by captaincaius** [comment]
Thanks for the response and sorry it wasn't clear.  I'll try to describe it in words.

tl;dr: I believe that if there's an optional property on a generic interface that's typed as a generic type variable, in cases it's not set, the type variable should collapse to "never", not "any".

The "why" relates to my example above.

On a generic interface with (1) an optional method and (2) a property, I'd like the property's type to be enforced as:
- If the optional method exists, the return type of the method.
- If the optional method doesn't exist, null.

But this seems impossible, because when the optional method doesn't exist, typescript leaves the generic type variable wide open.

My semantics might be a little rough above, but I hope this helps.

**Comment by RyanCavanaugh** [comment]
This isn't really an actionable description of a change to the current inference model. What's happening here is a zero-candidate inference that falls back to the constraint type. The missed optional properties don't really figure into it at all.
